gradle-nonnull-plugin/plugin/src/main/groovy/com.novoda.gradle.nonnull/AndroidNonNullPlugin.groovy

```diff
@@ -20,10 +24,81 @@ public class AndroidNonNullPlugin implements Plugin<Project> {
     private static void applyAndroid(project, variants) {
         variants.all { variant ->
 
-            def task = project.task("generate${variant.name.capitalize()}NonNullAnnotations", type: GeneratePackageAnnotationsTask)
-            task.outputDir = project.file("${project.buildDir}/generated/source/nonNull/${variant.dirName}")
-            task.variant = variant
+            def outputPath = "${project.buildDir}/generated/source/nonNull/${variant.dirName}"
+            def sourceSets = variant.sourceSets
+
+            def task = createTask(project, variant.name, outputPath, sourceSets)
+
             variant.registerJavaGeneratingTask(task, task.outputDir)
         }
     }
+
+    private static createTask(project, taskName, outputPath, sourceSets) {
+        def task = project.task("generate${taskName.capitalize()}NonNullAnnotations", type: GeneratePackageAnnotationsTask)
+        task.outputDir = project.file(outputPath)
+        task.sourceSets = sourceSets
+        task
+    }
+
+    private static void applyJava(project) {
+        configureTask(project)
+        configureIdeaModule(project)
+    }
+
+    private static void configureTask(project) {
+        project.sourceSets.all { sourceSet ->
+            String sourceSetName = (String) sourceSet.name
+            String taskName = "main".equals(sourceSetName) ? '' : sourceSetName
+
+            def generatedSourcesDir = "${project.buildDir}/generated/source/nonNull/${sourceSet.name}"
+            GeneratePackageAnnotationsTask task = createTask(project, taskName, generatedSourcesDir, [sourceSet])
+
+            Task classesTask = project.tasks.getByName(taskName.isEmpty() ? "classes" : "${taskName}Classes")
+            classesTask.mustRunAfter task
+
+            JavaCompile compileTask =
+                    (JavaCompile) project.tasks.getByName("compile${taskName.capitalize()}Java")
+            compileTask.source += project.fileTree(task.outputDir)
+            compileTask.dependsOn(task)
+
+        }
+    }
+
+    // inspired by https://github.com/tbroyer/gradle-apt-plugin/blob/master/src/main/groovy/net/ltgt/gradle/apt/AptPlugin.groovy#L171-L213
+    private static void configureIdeaModule(Project project) {
+        // so the user does not need to apply it when using the plugin
```

Comment:
I'm embarrassed. The screenshot was like that because I opened the project that already has the files. 

We tried with Eclipse with a brand new clone of the project. With or without `idea` plugin, we couldn't make it work.